src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -0,0 +1,436 @@
+        // WAIT_ABANDONED, WAIT_TIMEOUT).
+
+        // Successful wait on first object. When waiting for multiple objects the
+        // return value is (WaitSuccess + waitIndex).
+        internal static int WaitSuccess = 0;
+
+        // The specified object is a mutex object that was not released by the
+        // thread that owned the mutex object before the owning thread terminated.
+        // When waiting for multiple objects the return value is (WaitAbandoned +
+        // waitIndex).
+        internal static int WaitAbandoned = 0x80;
+
+        internal static int WaitTimeout = 0x102;
+
+        protected WaitHandle()
+        {
+        }
+
+        [Obsolete("Use the SafeWaitHandle property instead.")]
+        public virtual IntPtr Handle
+        {
+            get
+            {
+                return _waitHandle == null ? InvalidHandle : _waitHandle.DangerousGetHandle();
+            }
+            set
+            {
+                if (value == InvalidHandle)
+                {
+                    // This line leaks a handle.  However, it's currently
+                    // not perfectly clear what the right behavior is here 
+                    // anyways.  This preserves Everett behavior.  We should 
+                    // ideally do these things:
+                    // *) Expose a settable SafeHandle property on WaitHandle.
+                    // *) Expose a settable OwnsHandle property on SafeHandle.
+                    if (_waitHandle != null)
+                    {
+                        _waitHandle.SetHandleAsInvalid();
+                        _waitHandle = null;
+                    }
+                }
+                else
+                {
+                    _waitHandle = new SafeWaitHandle(value, true);
+                }
+            }
+        }
+
+        public SafeWaitHandle SafeWaitHandle
+        {
+            get
+            {
+                if (_waitHandle == null)
+                {
+                    _waitHandle = new SafeWaitHandle(InvalidHandle, false);
+                }
+                return _waitHandle;
+            }
+
+            set
+            { _waitHandle = value; }
+        }
+
+        internal static int ToTimeoutMilliseconds(TimeSpan timeout)
+        {
+            var timeoutMilliseconds = (long)timeout.TotalMilliseconds;
+            if (timeoutMilliseconds < -1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout), SR.ArgumentOutOfRange_NeedNonNegOrNegative1);
+            }
+            if (timeoutMilliseconds > int.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout), SR.ArgumentOutOfRange_LessEqualToIntegerMaxVal);
+            }
+            return (int)timeoutMilliseconds;
+        }
+
+        public virtual void Close() => Dispose();
+
+        protected virtual void Dispose(bool explicitDisposing)
+        {
+            _waitHandle?.Close();
```

Comment:
Should this be guarded behind `if (explicitDisposing)`?

Reply:
Seems unnecessary. It was not there before and it's a `SafeHandle`, so calling `Close` should be quite safe. I can add it if you insist.